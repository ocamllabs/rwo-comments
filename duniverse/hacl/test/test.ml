let pp_result ppf = function
  | Ok result -> Cstruct.hexdump_pp ppf result
  | Error e -> Format.fprintf ppf "error: %a@." Hacl_x25519.pp_error e

let test ~name ~pub ~priv =
  let result = Hacl_x25519.key_exchange priv pub in
  Format.printf "%s:@,%a@," name pp_result result

let key_pair_of_cstruct data = Hacl_x25519.gen_key ~rng:(fun _ -> data)

let key_pair_of_priv_hex s = key_pair_of_cstruct (Cstruct.of_hex s)

let too_short = Cstruct.create 31

let too_long = Cstruct.create 33

(** Test private-to-public conversion and key exchange.
    Data comes from RFC7748 6.1 and Wycheproof. *)
let run_tests () =
  let alice_private, alice_public =
    key_pair_of_priv_hex
      {| 77076d0a7318a57d3c16c17251b26645
         df4c2f87ebc0992ab177fba51db92c2a |}
  in
  test ~name:"public too short" ~priv:alice_private ~pub:too_short;
  test ~name:"public too long" ~priv:alice_private ~pub:too_long;
  Format.printf "alice_public:@.%a" Cstruct.hexdump_pp alice_public;
  let bob_private, bob_public =
    key_pair_of_priv_hex
      {| 5dab087e624a8a4b79e17f8b83800ee6
         6f3bb1292618b6fd1c2f8b27ff88e0eb |}
  in
  Format.printf "bob_public:@.%a" Cstruct.hexdump_pp bob_public;
  test ~name:"pub_a * priv_b" ~pub:alice_public ~priv:bob_private;
  test ~name:"pub_b * priv_a" ~pub:bob_public ~priv:alice_private;
  let zeroes = Cstruct.create 32 in
  test ~name:"pub = 0" ~pub:zeroes ~priv:alice_private;
  let low_order_pub =
    Cstruct.of_hex
      {| e0eb7a7c3b41b8ae1656e3faf19fc46a
         da098deb9c32b1fd866205165f49b800 |}
  in
  let low_order_priv, _ =
    key_pair_of_priv_hex
      {| 10255c9230a97a30a458ca284a629669
         293a31890cda9d147febc7d1e22d6bb1 |}
  in
  test ~name:"low order point" ~pub:low_order_pub ~priv:low_order_priv

let () = run_tests ()
