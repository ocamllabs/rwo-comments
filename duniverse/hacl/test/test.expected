public too short:
error: Invalid key size
public too long:
error: Invalid key size

alice_public:
85 20 f0 09 89 30 a7 54  74 8b 7d dc b4 3e f7 5a
0d bf 3a 0d 26 38 1a f4  eb a4 a9 8e aa 9b 4e 6a
bob_public:
de 9e db 7d 7b 7d c1 b4  d3 5b 61 c2 ec e4 35 37
3f 83 43 c8 5b 78 67 4d  ad fc 7e 14 6f 88 2b 4f
pub_a * priv_b:
4a 5d 9d 5b a4 ce 2d e1  72 8e 3b f4 80 35 0f 25
e0 7e 21 c9 47 d1 9e 33  76 f0 9b 3c 1e 16 17 42
pub_b * priv_a:
4a 5d 9d 5b a4 ce 2d e1  72 8e 3b f4 80 35 0f 25
e0 7e 21 c9 47 d1 9e 33  76 f0 9b 3c 1e 16 17 42
pub = 0:
error: Public key with low order
low order point:
error: Public key with low order

